spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+in place for dealing with such behavior is `ProcessProposal`.
+
+##### Replay Protection
+
+To prevent old transactions from being replayed, `CheckTx` must implement
+replay protection.
+
+It is possible for old transactions to be sent to the application. So
+it is important `CheckTx` implements some logic to handle them.
+
+#### Info/Query Connection
+
+The Info (or Query) Connection should maintain a `QueryState` for answering queries from the user,
+and for initialization when Tendermint first starts up (both described further
+below).
+
+`QueryState` always contain a copy of of the latest `ExecuteTxState` at the end of every `Commit`,
+after the full block has been processed and the state committed to disk.
+Otherwise it should never be modified.
+
+Tendermint Core currently uses the Query connection to filter peers upon
+connecting, according to IP address or node ID. For instance,
+returning non-OK ABCI response to either of the following queries will
+cause Tendermint to not connect to the corresponding peer:
+
+* `p2p/filter/addr/<ip addr>`, where `<ip addr>` is an IP address.
+* `p2p/filter/id/<id>`, where `<is>` is the hex-encoded node ID (the hash of
+  the node's p2p pubkey).
+
+Note: these query formats are subject to change!
+
+#### Snapshot Connection
+
+The Snapshot Connection is optional, and is only used to serve state sync snapshots for other nodes
+and/or restore state sync snapshots to a local node being bootstrapped.
+
+For more information, see [the state sync section](#state-sync) below.
+
+### Transaction Results
+
+The Application is expected to return a list of
+[`ExecTxResult`](./abci%2B%2B_methods_002_draft.md#exectxresult) in
+[`ResponseFinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock). The list of transaction
+results must respect the same order as the list of transactions delivered via
+[`RequestFinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock).
+This section discusses the fields inside this structure, along with the fields in
+[`ResponseCheckTx`](./abci%2B%2B_methods_002_draft.md#checktx),
+whose semantics are similar.
+
+The `Info` and `Log` fields are
+non-deterministic values for debugging/convenience purposes that are otherwise ignored.
+
+#### Gas
+
+Ethereum introduced the notion of *gas* as an abstract representation of the
+cost of resources used by nodes when processing transactions. Every operation in the
+Ethereum Virtual Machine uses some amount of gas, and gas can be accepted at a market-variable price.
+Users propose a maximum amount of gas for their transaction; if the transaction uses less, they get
+the difference credited back. Tendermint adopts a similar abstraction,
+though uses it only optionally and weakly, allowing applications to define
+their own sense of the cost of execution.
+
+In Tendermint, the [ConsensusParams.Block.MaxGas](../proto/types/params.proto) limits the amount of gas
+that can be used in a block.
```

Comment:
More specifically, for the transactions included in a block? Does it also count for something else in a block?

Reply:
I don't think so. I'll clarify